### PR TITLE
Fix Incorrect Metric Reporting Post-Update in Terraform Controller

### PR DESCRIPTION
**Pull request by TarasLykhenko**
Fixes: #1068

This PR addresses a bug in the Tofu Controller where the gotk_reconcile_condition metrics for the Ready condition were not being accurately reported, causing the metrics to perpetually show as updating even after a successful update.

Issue:
Post-update, the metrics incorrectly reported Unknown=1 with True/False=0, due to the misuse of the original terraform variable in a deferred function. This variable held the Unknown condition and inadvertently overwrote the correctly updated metrics.

Root Cause:
The defer statement used the original terraform variable, which contained the Unknown condition. This action overwrote the metrics that were updated correctly elsewhere in the code at a different point ([see here](https://github.com/weaveworks/tf-controller/blob/1958307fc83cc56f486453885e9426cd88f93d80/controllers/tf_controller.go#L439)) using the reconciledTerraform variable.

Resolution:
The fix involves modifying the defer statement so it no longer updates the metric. This ensures that the metric updates performed through the reconciliation code in r.RecordReadiness are preserved and not overwritten by outdated data.

By merging this PR, the metrics accurately reflect the current state of Terraforms, thereby improving our system's reliability of status reporting.

### CODE REVIEW

File: controllers/tf_controller.go [hunk 1]
```diff
@@ -136,7 +136,6 @@ func (r *TerraformReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		}
 
 		// Record Prometheus metrics.
-		r.Metrics.RecordReadiness(ctx, &terraform)
 		r.Metrics.RecordSuspend(ctx, &terraform, terraform.Spec.Suspend)
 		r.Metrics.RecordDuration(ctx, &terraform, reconcileStart)
```
Comment (yitsushi) [inline controllers/tf_controller.go:140]: Are these safe from this? They care using the same reference, or are those functions check extra criteria that wasn't checked in `RecordReadiness` so they are not affected by the same issue?

Comment (TarasLykhenko) [reply controllers/tf_controller.go:140]: Yes, these functions check extra criteria that aren't checked in the RecordReadiness, so they are not affected by the overwrite issue.

**Comment by TarasLykhenko** [comment]
@yitsushi, the e2e test is failing, but it seems unrelated to this change. What needs to be done to unblock it?

**Comment by ilithanos** [comment]
@TarasLykhenko #1290 should fix the e2e tests when merged.

**Comment by ricardomiguel-os** [comment]
Hi @yitsushi , is it possible to merge this change? thank you

**Comment by yitsushi** [comment]
Ah the rebase/merge invalidated the approval 🤦‍♀️

**Review by yitsushi — APPROVED** [review]

**Review by chanwit — APPROVED** [review]
LGTM :1st_place_medal: 

Thank you @TarasLykhenko for the fix!

**Comment by ricardomiguel-os** [comment]
@yitsushi, this action failed to push the image.

https://github.com/flux-iac/tofu-controller/actions/runs/8949941026/job/24584776720

Do you think that a retry might solve the issue? Thanks

**Comment by yitsushi** [comment]
> @yitsushi, this action failed to push the image.
> 
> https://github.com/flux-iac/tofu-controller/actions/runs/8949941026/job/24584776720
> 
> Do you think that a retry might solve the issue? Thanks

hmm, it seems it should, unless they removed `alpine:3.19` which doesn't seem reasonable.

> ERROR: failed to solve: alpine:3.19: failed to resolve source metadata for docker.io/library/alpine:3.19: failed to do request: Head "https://registry-1.docker.io/v2/library/alpine/manifests/3.19": EOF

**Comment by ricardomiguel-os** [comment]
@yitsushi, do you know when the next RC version will be released?

I'm just asking because version v.0.14.1 had some critical vulnerabilities, and the current one has this metrics bug. Thanks

**Comment by yitsushi** [comment]
No, I don't know.

**Pull request merged**